### ISSUE TITLE
BlockingChannel can't stop_consuming

### ISSUE THREAD

**Issue by antoineleclair** [pika/pika#770]
When using a `BlockingChannel` to consume many queues with `start_consuming`, there's no way to `stop_consuming` cleanly, unless a message is received. 

```python
import signal
import pika

params = pika.URLParameters('your params')
connection = pika.BlockingConnection(params)
channel = connection.channel()
channel.queue_declare(queue='queue_one', durable=True)
channel.queue_declare(queue='queue_two', durable=True)
channel.basic_qos(prefetch_count=1)

def callback(ch, method, properties, body):
    # process message here
    ch.basic_ack(delivery_tag=method.delivery_tag)

channel.basic_consume(callback, queue=queue)

def sigterm_received(signum, frame):
    channel.stop_consuming()
signal.signal(signal.SIGTERM, sigterm_received)

channel.start_consuming()

print('Never reaches here unless a message is dequeued')

channel.close()
connection.close()
```

If a message is received, then the message "Never reaches hhere unless a message is dequeued" is printed. My best guess is that this line is causing the problem:

```python
while self._consumer_infos:
    self.connection.process_data_events(time_limit=None)
```

https://github.com/pika/pika/blob/51fd186f1fabe0da8582609d266df1e28e57e3f4/pika/adapters/blocking_connection.py#L1755

In the example above, replacing the line

```python
channel.start_consuming()
```

with

```python
while channel._consumer_infos:
    channel.connection.process_data_events(time_limit=1)
```

seems to fix the issue. However, I'm a bit cautious that this fix will cause other issues due to the fact that the `BlockingConnection` is not thread safe.

Also, just to clarify what I'm trying to do here: I want to consume many queues and shutdown gracefully when a SIGTERM signal is received. I want the messages that are currently processing to complete (and ACK) before closing everything. And I also want to exit quick (1 second is acceptable) if there are no messages currently being processed.

**Comment by vitaly-krugl** [comment]
@antoineleclair, I would never call something from a signal handler that's not explicitly documented as "signal-safe". That's asking for trouble. Also, using the private instance variable `_consumer_infos` is problematic, because it may change in the future.

The following may work for you, using documented public API:

1. In `sigterm_received`, set a global variable to True (e.g., `g_stop_requested = True`) or something that's equivalent and signal-safe.
2. Use `BlockingConnection.add_timeout` to request a 1-second timer callback
3. In the callback, check if your `g_stop_requested` (or equivalent) is set, and call `BlockingChannel.stop_consuming` to cancel all consumers, if so. If not, request a new timer callback.

I am generally not crazy about polling, but `BlockingConnection` implements the synchronous command metaphor, so there is not much else you can do, I think. Also, calling `channel.connection.process_data_events(time_limit=1)` is also polling and using a timer under the covers.

Hope this helps!

**Comment by vitaly-krugl** [comment]
Please close the issue if my reply resolved the problem.

**Comment by vitaly-krugl** [comment]
@antoineleclair - And another thing, if there is the possibility that your RabbitMQ server may run low on resources (low disk space, low RAM), then RabbitMQ will go into blocking mode (won't be reading from your connection), which may block your `basic_ack` or `basic_publish` call indefinitely. In that case, the timer callbacks won't be processed until RabbitMQ becomes unblocked.

For that, there is a new feature in master branch (not released yet), that allows you to set a blocked connection timeout. Find out more here: https://github.com/pika/pika/blob/7922e076c2b5ded29156b9523dc5cf8178c3f1ce/pika/adapters/blocking_connection.py#L285-L307

**Comment by antoineleclair** [comment]
Thanks @vitaly-krugl, that did it. Much cleaner with `BlockingConnection.add_timeout`. Thanks a lot!

Also, if my opinion can be useful, I would have expected, from an API design perspective, `BlockingChannel.stop_consuming` to be signal-safe and to work even when no messages are being received.

**Comment by vitaly-krugl** [comment]
@antoineleclair, thank you for your update and feedback.

A pika connection is not thread-safe. Same concerning signal-safety. Supporting those paradigms typically introduces significant complexity and performance issues and is outside the scope of this library.

**Comment by lukebakken** [comment]
@guysoft - please open a new issue with a description of what you are trying to accomplish and what you observe. Also, provide a runnable code sample that demonstrates it. That will be the most help to Pika's maintainers, who are all volunteers. Thank you!

**Comment by guysoft** [comment]
Ok, I fixed my issue (using basic_consume and handling it myself). But basically the issue was this:
When running the add_timeout callback you can't interact with part of the python enviroment, even though its same thread.
I can't manage to rip example code, I might have time to solve this later on. Thanks for the reply though and well done!

**Comment by wanguanfu** [comment]
I throw no clear way to solve （pika-stop-consuming-does-not-work）

**Comment by wanguanfu** [comment]
like this？？？

       In the example above, replacing the line
                 channel.start_consuming()
   with
                 while channel._consumer_infos:
                           channel.connection.process_data_events(time_limit=1)

**Comment by lukebakken** [comment]
@wanguanfu - please provide code that reproduces your issue and attach it as a new message to the [`pika-python`](https://groups.google.com/forum/#!forum/pika-python) mailing list.
